File: exercise/12-integer-to-roman/Mister-Hope.cpp
/*
 * Runtime: 24 ms, faster than 22.26% of C++ online submissions for Integer to
 * Roman.
 *
 * Memory Usage: 12.6 MB, less than 11.00% of C++ online submissions for Integer
 * to Roman.
 */

#include <iostream>
#include <vector>
using namespace std;
class Solution {
 public:
  string intToRoman(int num) {
    if (num <= 0) return "";

    vector<string> thousands{"", "M", "MM", "MMM"};
    vector<string> hundreds{"",  "C",  "CC",  "CCC",  "CD",
                            "D", "DC", "DCC", "DCCC", "CM"};
    vector<string> tens{"",  "X",  "XX",  "XXX",  "XL",
                        "L", "LX", "LXX", "LXXX", "XC"};
    vector<string> units{"",  "I",  "II",  "III",  "IV",
                         "V", "VI", "VII", "VIII", "IX"};

    return thousands[num / 1000] + hundreds[(num % 1000) / 100] +
           tens[(num % 100) / 10] + units[num % 10];
  }
};

File: leetcode/12-integer-to-roman/Mister-Hope.js
/*
 * Runtime: 176 ms, faster than 33.68% of JavaScript online submissions for Integer to Roman.
 *
 * Memory Usage: 45.2 MB, less than 6.38% of JavaScript online submissions for Integer to Roman.
 */

const intToRoman = (num) => {
  const numerals = [
    "M",
    "CM",
    "D",
    "CD",
    "C",
    "XC",
    "L",
    "XL",
    "X",
    "IX",
    "V",
    "IV",
    "I",
  ];
  const weights = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];

  let result = "";

  for (let i = 0; i < weights.length && num !== 0; i++) {
    for (let j = 1; j <= num / weights[i]; j++) result += numerals[i];

    num %= weights[i];
  }

  return result;
};
